File: apps/website/src/server/db/seeds/corte-de-mayor-ganancia/solution-dynamicProgramming.py
def corteDeMayorGanancia(precios, longitud):
    memo = [None for i in range(longitud + 1)]
    return aux(precios, memo, longitud)


def aux(precios, memo, longitudActual):
    if longitudActual == 0:
        return 0
    if memo[longitudActual] is None:
        memo[longitudActual] = 0
        for i in range(1, longitudActual + 1):
            indiceReal = i - 1
            memo[longitudActual] = max(
                memo[longitudActual],
                precios[indiceReal] + aux(precios, memo, longitudActual - i),
            )
    return memo[longitudActual]

File: apps/website/src/server/db/seeds/corte-de-mayor-ganancia/solution-dynamicProgramming.js

function corteDeMayorGanancia(precios, longitud) {
  const memo = Array(longitud + 1);

  function aux(longitudActual) {
    if (longitudActual === 0) {
      return 0;
    }
    if (memo[longitudActual] === undefined) {
      memo[longitudActual] = 0;
      for (let i = 1; i <= longitudActual; i++) {
        const indiceReal = i - 1;
        memo[longitudActual] = Math.max(
          memo[longitudActual],
          precios[indiceReal] + aux(longitudActual - i),
        );
      }
    }

    return memo[longitudActual];
  }
  return aux(longitud);
}
